aiida_optimize/engines/_convergence.py

```diff
@@ -0,0 +1,236 @@
+        if not self.initialized:
+            return False
+
+        # calculate pair distances between results
+        distance_triangle = self._distance_triangle
+        # flatten all the distances into a 1D list
+        distances = list(itertools.chain(*distance_triangle))
+
+        # check if the maximum distance is less than the tolerance
+        return np.max(distances) < self.tol
+
+    @property
+    def is_finished(self) -> bool:
+        if not self.initialized:
+            return False
+
+        # If we've used all the input values, we're finished
+        if len(self.result_values) >= len(self.input_values):
+            return True
+
+        # If _num_new_iters is -1, we know that we cannot converge with
+        # the remaining inputs in input_values, so we're finished
+        if self._num_new_iters == -1:
+            return True
+
+        return self.is_converged
+
+    @property
+    def is_finished_ok(self) -> bool:
+        if self.is_finished and self.is_converged:
+            return True
+        return False
+
+    def _create_inputs(self) -> ty.List[ty.Dict[str, orm.Float]]:
+        """
+        Create the inputs for the evaluation function.
+        If the work chain is not initialized, the appropriate number of
+        inputs to fill the convergence window are generated.
+        Otherwise, one more input will be generated
+        """
+        if not self.initialized:
+            # Do enough calculations to fill the initial convergence window
+            num_new_iters = self.convergence_window
+            self.initialized = True
+        else:
+            num_new_iters = self._num_new_iters
+
+        self.current_index += num_new_iters
+        inputs = [{
+            self.input_key: orm.Float(self.input_values[i])
+        } for i in range(self.current_index - num_new_iters, self.current_index)]
+
+        return inputs
+
+    def _update(self, outputs: ty.Dict[int, ty.Any]) -> None:
+        """
+        Update the state of the engine by saving all the output values and
+        storing them in the result_values property
+        """
+        output_values = outputs.values()  # Don't need keys (AiiDA UUIDs)
+        self.result_values += [
+            val[self.result_key] for val in output_values
+        ]  # Values are AiiDA types
+
+    def _get_optimal_result(self) -> ty.Tuple[int, Result]:
+        """
+        Retrieve the converged index and result value (output value, _not_ max
+        distance within convergence window)
+        """
+        return (
+            len(self.result_values) - self.convergence_window,
+            self.result_values[-self.convergence_window]
+        )
+
+    @property
+    def result_value(self) -> Result:
+        return self._get_optimal_result()[1]
+
+    @property
+    def result_index(self) -> int:
+        return self._get_optimal_result()[0]
+
+
+class Convergence(OptimizationEngineWrapper):
+    """
+    Wrapper class for convergence engine
```

Comment:
The engine-specific inputs should be documented here.

Reply:
What documentation style should I follow? I didn't put any documentation because the other engines do not have any.

Reply:
They do - it's on the "wrapper", not the implementation, e.g. https://github.com/greschd/aiida-optimize/blob/develop/aiida_optimize/engines/_bisection.py#L115

The existing engines use "plain" restructuredtext, but TBH I much prefer numpy style - let's just use that, and I'll switch the other ones over later.

Reply:
Ah my bad, I'll put in numpy style docs then.

Reply:
Can you add `sphinx.ext.napoleon` in the `conf.py`? https://github.com/greschd/aiida-optimize/blob/develop/doc/source/conf.py#L41

Reply:
Since I was a bit late with that comment, I added it myself.